contracts/Committee.sol

```diff
@@ -166,374 +233,479 @@ contract Committee is ICommittee, Ownable {
-			newCommitteeSize++;
-		}
-		committeeSize = newCommitteeSize;
-		_refreshReadyForCommitteeCount();
-		_refreshOldestReadyToSyncStandbyPos();
-		return (prevCommitteeSize, newCommitteeSize);
+	function _rankAndUpdateMember(Member memory member) private returns (bool committeeChanged, bool standbysChanged) {
+		(committeeChanged, standbysChanged) = _rankMember(member);
+		membersData[member.addr] = member.data;
 	}
 
-	function _refreshReadyForCommitteeCount() private returns (uint, uint) {
-		uint newCount = readyForCommitteeCount;
-		uint prevCount = newCount;
-		while (newCount > 0 && (topology.length < newCount || !isReadyForCommittee(topology[newCount - 1]))) {
-			newCount--;
-		}
-		while (topology.length > newCount && isReadyForCommittee(topology[newCount])) {
-			newCount++;
-		}
-		readyForCommitteeCount = newCount;
-		return (prevCount, newCount);
-	}
-
-	function _refreshOldestReadyToSyncStandbyPos() private {
-		uint256 oldestTimestamp = uint(-1);
-		uint oldestPos = uint(-1);
-		for (uint i = committeeSize; i < topology.length; i++) {
-			uint t = members[topology[i]].readyToSyncTimestamp;
-			if (t < oldestTimestamp) {
-				oldestTimestamp = t;
-				oldestPos = i;
+	function _rankMember(Member memory member) private returns (bool committeeChanged, bool standbysChanged) {
+		Settings memory _settings = settings;
+
+		bool isParticipant = member.data.inCommittee || member.data.isStandby;
+		if (!isParticipant) {
+			CommitteeInfo memory _committeeInfo = committeeInfo;
+			if (!qualifiedToJoin(member, _settings, _committeeInfo)) {
+				return (false, false);
 			}
+			joinMemberAsParticipant(member, _committeeInfo);
 		}
-		oldestReadyToSyncStandbyPos = int(oldestPos);
-	}
 
-	function _compareValidatorsByData(address v1, uint256 v1Weight, bool v1Ready, address v2, uint256 v2Weight, bool v2Ready) private pure returns (int) {
-		return v1Ready && !v2Ready ||
-				v1Ready == v2Ready  && v1Weight > v2Weight ||
-				v1Ready == v2Ready  && v1Weight == v2Weight && uint256(v1) > uint256(v2)
-		? int(1) : -1;
+		return updateOnMemberChange(member, _settings);
 	}
 
-	function _compareValidators(address v1, address v2) private view returns (int) {
-		bool v1Ready = isReadyForCommittee(v1);
-		uint256 v1Weight = getValidatorWeight(v1);
-		bool v2Ready = isReadyForCommittee(v2);
-		uint256 v2Weight = getValidatorWeight(v2);
-		return _compareValidatorsByData(v1, v1Weight, v1Ready, v2, v2Weight, v2Ready);
+	function qualifiedToJoin(Member memory member, Settings memory _settings, CommitteeInfo memory _committeeInfo) private view returns (bool) {
```

Comment:
Agree, although the way it's written now may be just a bit more readable.

Reply:
my problem is with the two very similar names `qualifiesForCommittee ` or `qualifiedToJoinCommittee ` which don't indicate the difference between them.

some ideas: 
- rename them to have the same name (function overloading - meh :/),  
- OR, rename one to express the different but use the same tense in both methods.

in addition - `qualifiedToJoin` what? it's `participant`s in the current terminology (see general commnet on the PR). if you keep the term Participant shouldnt `qualifiedToJoin ` read `qualifiesForParticipation` or `qualifiesToParticipate`?

Reply:
Yes, I agree `qualifiesToParticipate` is more consistent with the terminology. My way of thinking when naming these is that `qualifiedToJoin` is for a member who is currently not in the committee, while `qualifiedForCommittee` is more general and can be applied for members who are currently in the committee. Anyway, let's go with your preference here because I can go either way. 

Do you prefer the variation in your original comment?

Reply:
in this case maybe change `qualifiedToJoin ` to `shouldJoinXXX` but again, better to use `Participants` or `Committee`. 

BTW - there is no such thing as `qualifiedForCommittee` which makes sense since disqualification happens implicitly..

Reply:
And regarding my original comment. I think "member" is not the best choice because its really a roundabout way to say "a validator i know of".

and participants is less descriptive than "standbysAndCommittee" but if you prefer that for brevity we can maybe state the definition of the term above the array declaration in a comment. as you prefer

Reply:
I'd replace member to some expression containing "validator" because member is just too arbitrary imho

Reply:
The "member" notation originates from the interface (addMember/removeMember/etc.). I don't mind changing it.

Reply:
opened #30